File: python/palindrome-products/palindrome_products.py
def largest(min_factor, max_factor):
    """Given a range of numbers, find the largest palindromes which
       are products of two numbers within that range.

    :param min_factor: int with a default value of 0
    :param max_factor: int
    :return: tuple of (palindrome, iterable).
             Iterable should contain both factors of the palindrome in an arbitrary order.
    """

    if max_factor < min_factor:
        raise ValueError("min must be <= max")

    value = None
    factors = []

    for a in range(max_factor, min_factor - 1, -1):
        if value is not None and value > a * max_factor:
            break

        for b in range(max_factor, a - 1, -1):
            if is_palindrome(a * b):
                if value is None or value < a * b:
                    value = a * b
                    factors = [[a, b]]
                elif value == a * b:
                    factors.append([a, b])
                else:
                    break

    return (value, factors)


def smallest(min_factor, max_factor):
    """Given a range of numbers, find the smallest palindromes which
    are products of two numbers within that range.

    :param min_factor: int with a default value of 0
    :param max_factor: int
    :return: tuple of (palindrome, iterable).
    Iterable should contain both factors of the palindrome in an arbitrary order.
    """

    if max_factor < min_factor:
        raise ValueError("min must be <= max")

    value = None
    factors = []

    for a in range(min_factor, max_factor + 1):
        if value is not None and value < a * a:
            break

        for b in range(a, max_factor + 1):
            if is_palindrome(a * b):
                if value is None or value > a * b:
                    value = a * b
                    factors = [[a, b]]
                elif value == a * b:
                    factors.append([a, b])
                else:
                    break

    return (value, factors)

def is_palindrome(number):
    if number % 10 == 0:
        return False

    rev = 0

    while number > rev:
        rev = rev * 10 + number % 10
        number //= 10

    return number == rev or number == rev // 10

File: javascript/palindrome-products/palindrome-products.js
export class Palindromes {
  static isPalindrome(num) {
    const str = num.toString();
    return str === str.split("").reverse().join("");
  }

  static generate(min, max) {
    let smallest = Infinity;
    let largest = -Infinity;
    const products = {};

    for (let i = min; i <= max; i++) {
      for (let j = i; j <= max; j++) {
        const product = i * j;
        if (this.isPalindrome(product)) {
          if (product < smallest) {
            smallest = product;
            product[smallest] = [[i, j]];
          } else if (product === smallest) {
            products[smallest].push([i, j]);
          }

          if (product > largest) {
            largest = product;
            products[largest] = [[i, j]];
          } else if (product === largest) {
            products[largest].push([i, j]);
          }
        }
      }
    }

    return { smallest, largest, factorPairs: products };
  }
}
